libraries/AP_Quicktune/AP_Quicktune.cpp

```diff
@@ -0,0 +1,690 @@
+    auto *pid = get_axis_pid(axis);
+    if (pid == nullptr) {
+        INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
+        return nullptr;
+    }
+    const Param roll_param = Param(uint8_t(param) % param_per_axis);
+    switch (roll_param)
+    {
+        case Param::RLL_P:
+            return &pid->kP();
+        case Param::RLL_I:
+            return &pid->kI();
+        case Param::RLL_D:
+            return &pid->kD();
+        case Param::RLL_SMAX:
+            return &pid->slew_limit();
+        case Param::RLL_FLTT:
+            return &pid->filt_T_hz();
+        case Param::RLL_FLTD:
+            return &pid->filt_D_hz();
+        case Param::RLL_FLTE:
+            return &pid->filt_E_hz();
+        case Param::RLL_FF:
+            return &pid->ff();
+        default:
+            INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
+            return nullptr;
+    }
+}
+
+float AP_Quicktune::get_param_value(AP_Quicktune::Param param) const
+{
+    AP_Float *ptr = get_param_pointer(param);
+    if (ptr != nullptr) {
+        return ptr->get();
+    }
+    return 0.0;
+}
+
+void AP_Quicktune::set_param_value(AP_Quicktune::Param param, float value)
+{
+    AP_Float *ptr = get_param_pointer(param);
+    if (ptr != nullptr) {
+       ptr->set(value);
+    }
+}
+
+void AP_Quicktune::set_and_save_param_value(AP_Quicktune::Param param, float value)
+{
+    AP_Float *ptr = get_param_pointer(param);
+    if (ptr != nullptr) {
+       ptr->set_and_save(value);
+    }
+}
+
+AP_Quicktune::AxisName AP_Quicktune::get_axis(AP_Quicktune::Param param) const
+{
+    if (param >= Param::END) {
+        return AxisName::END;
+    }
+    return AxisName(uint8_t(param) / param_per_axis);
+}
+
+const char* AP_Quicktune::get_axis_name(AP_Quicktune::AxisName axis) const
+{
+    switch (axis)
+    {
+        case AxisName::RLL:
+            return "Roll";
+        case AxisName::PIT:
+            return "Pitch";
+        case AxisName::YAW:
+            return "Yaw";
+        default:
+            INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
+            return "UNK";
+    }
+}
+
+float AP_Quicktune::gain_limit(AP_Quicktune::Param param) const
+{
+    if (get_axis(param) == AxisName::YAW) {
+        if (param == Param::YAW_P) {
+            return yaw_p_max;
+        }
+        if (param == Param::YAW_D) {
+            return yaw_d_max;
+        }
+    }
+    return 0.0;
+}
+
+
+// @LoggerMessage: QUIK
+// @Description: Quicktune
+// @Field: TimeUS: Time since system startup
+// @Field: SRate: slew rate
+// @Field: Gain: test gain for current axis and PID element
+// @Field: Param: name of parameter being being tuned
+// @Field: ParamNo: number of parameter being tuned
+void AP_Quicktune::Write_QUIK(float srate, float gain, AP_Quicktune::Param param)
+{
+#if HAL_LOGGING_ENABLED
+    AP::logger().WriteStreaming("QUIK","TimeUS,SRate,Gain,Param,ParamNo", "QffNI",
```

Comment:
If you add units you could mark "ParamNo" as a instance which would make it much nice in log review.

Reply:
i've added that, it does result in a bit odd looking log as the indexes are a bit unintuitive, but overall an improvement I think